### PR TITLE
fixes: audit snarkpack aggregation

### PR DESCRIPTION
**Pull request by dignifiedquire**
for details see commits

### CODE REVIEW

File: src/groth16/aggregate/prove.rs [hunk 1]
```diff
@@ -257,25 +277,31 @@ fn gipa_tipp_mipp<E: Engine>(
...
-        // Optimization for multiexponentiation to rescale G2 elements with
-        // 128-bit challenge Swap 'c' and 'c_inv' since can't control bit size
-        // of c_inv
-        let c = c_inv.inverse().unwrap();
+        if first_loop {
+            first_loop = false;
+        // already generated c_inv and c outside of the loop
+        } else {
+            c_inv = *Transcript::<E>::new("gipa")
+                .write(&c_inv)
```
Comment (nikkolasg) [inline src/groth16/aggregate/prove.rs:285]: to be fair I think we should hash the c and not c_inv - doesnt make much sense to hash the inverse

**Review by nikkolasg — APPROVED** [review]

**Comment by arajasek** [comment]
<3

File: src/groth16/aggregate/prove.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/groth16/aggregate/prove.rs:284]: well the paper says to hash the output of the previous hash, which is why I used `c_inv`

**Review by nikkolasg — APPROVED** [review]

**Comment by dignifiedquire** [comment]
reviewed with @nicola live

**Pull request merged**